File: Builder/UnitTests/ExtensionMethods.cs
﻿
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Xml.Linq;
using System.Xml.Serialization;
using Cito.Tester.Common;
using Cito.Tester.ContentModel;
using Questify.Builder.Logic.ContentModel.Scoring;
using Questify.Builder.Model.ContentModel.EntityClasses;

namespace Questify.Builder.UnitTests
{
    public static class ExtensionMethods
    {

        /// <summary>
        /// Sets the Assessment on ItemResourceEntity.
        /// </summary>
        /// <param name="me"></param>
        /// <param name="a"></param>
        public static void SetAssessmentOnItemResource(this ItemResourceEntity me, AssessmentItem a)
        {
            if (me.ResourceData == null) me.ResourceData = new ResourceDataEntity();

            using (var stream = new MemoryStream())
            {
                SerializeHelper.XmlSerializeToStream(stream, a);
                me.ResourceData.BinData = stream.ToArray();
            }
        }


        /// <summary>
        /// Gets the Assessment form ItemResourceEntity
        /// </summary>
        /// <param name="me"></param>
        public static AssessmentItem GetAssessmentFromItemResource(this ItemResourceEntity me)
        {
            //guards
            if (me.ResourceData == null) return null;
            if (me.ResourceData.BinData.Length == 0) return null;

            return (AssessmentItem)SerializeHelper.XmlDeserializeFromByteArray(me.ResourceData.BinData, typeof(AssessmentItem));
        }

        /// <summary>
        /// Transforms ScoringParameter to a CombinedScoringMapKey
        /// </summary>
        /// <param name="me">Me.</param>
        /// <returns>CombinedScoringMapKey.</returns>
        public static CombinedScoringMapKey AsCombinedScoringMap(this ScoringParameter me)
        {
            List<ScoringMapKey> keys = new List<ScoringMapKey>();
            foreach (var set in me.Value)
            {
                keys.Add(new ScoringMapKey(me, set.Id));
            }

            return CombinedScoringMapKey.Create(keys);
        }

        /// <summary>
        /// Transforms ScoringParameter to a CombinedScoringMapKey
        /// </summary>
        /// <param name="me">Me.</param>
        /// <param name="groupIdentifiers"></param>
        /// <returns>CombinedScoringMapKey.</returns>
        public static CombinedScoringMapKey AsCombinedScoringMapThatIsGroup(this ScoringParameter me, IEnumerable<int> groupIdentifiers )
        {
            List<ScoringMapKey> keys = new List<ScoringMapKey>();
            foreach (var set in me.Value)
            {
                keys.Add(new ScoringMapKey(me, set.Id));
            }

            return CombinedScoringMapKey.Create(keys, groupIdentifiers);
        }

        /// <summary>
        /// Transforms ScoringParameter to a CombinedScoringMapKey
        /// </summary>
        /// <param name="me">Me.</param>
        /// <param name="solution"></param>
        /// <returns>CombinedScoringMapKey.</returns>
        public static CombinedScoringMapKey AsCombinedScoringMap(this ScoringParameter me, Solution solution)
        {
            List<ScoringParameter> sp = new List<ScoringParameter>() { me };
            return new ScoringMap(sp, solution).GetMap().First();
        }

        /// <summary>
        /// Transforms ScoringParameter to a CombinedScoringMapKey with factSets.
        /// </summary>
        /// <param name="me">Me.</param>
        /// <param name="setNumbers">The set numbers.</param>
        /// <returns>CombinedScoringMapKey.</returns>
        public static CombinedScoringMapKey AsCombinedScoringMap(this ScoringParameter me, params int[] setNumbers)
        {
            List<ScoringMapKey> keys = new List<ScoringMapKey>();
            foreach (var set in me.Value)
            {
                keys.Add(new ScoringMapKey(me, set.Id));
            }
            return CombinedScoringMapKey.Create(keys, setNumbers);
        }

        /// <summary>
        /// Adds the sub parameters.
        /// </summary>
        /// <typeparam name="TScorePrm">The type of the t score PRM.</typeparam>
        /// <param name="me">Me.</param>
        /// <param name="ids">The ids.</param>
        /// <returns>TScorePrm.</returns>
        public static TScorePrm AddSubParameters<TScorePrm>(this TScorePrm me, params string[] ids)
            where TScorePrm : ScoringParameter
        {
            me.Value = new ParameterSetCollection();
            foreach (var id in ids)
                me.Value.Add(new ParameterCollection() { Id = id });
            return me;
        }

        public static TScorePrm AddGapTextParameters<TScorePrm>(this TScorePrm me)
            where TScorePrm : ScoringParameter
        {
            foreach (var collection in me.Value)
            {
                collection.InnerParameters.Add(new GapTextParameter() { Value = collection.Id });
            }
            return me;
        }

        public static XElement DoSerialize<T>(this T obj)
        {
            var s = new XmlSerializer(typeof(T));
            XElement ret = null;
            using (StringWriter m = new StringWriter())
            {
                s.Serialize(m, obj);
                ret = XElement.Parse(m.ToString());
            }
            return ret;
        }

        public static T To<T>(this XElement input)
        {
            T ret = default(T);
            var s = new XmlSerializer(typeof(T));

            using (StringReader m = new StringReader(input.ToString()))
            {
                ret = (T)s.Deserialize(m);
            }

            return ret;
        }

        public static T To<T>(this byte[] input)
        {
            T ret = default(T);
            var s = new XmlSerializer(typeof(T));

            using (MemoryStream m = new MemoryStream(input))
            {
                ret = (T)s.Deserialize(m);
            }

            return ret;
        }

        public static byte[] ToBytes(this XElement input)
        {
            return Encoding.UTF8.GetBytes(input.ToString());
        }

        public static string AsString(this  byte[] input)
        {
            return Encoding.UTF8.GetString(input);
        }

    }
}

File: Builder/UnitTests.Framework/ExtensionMethods.vb
Imports System.Runtime.CompilerServices
Imports System.Xml.Linq
Imports System.Linq
Imports System.Xml
Imports System.IO
Imports Questify.Builder.Model.ContentModel.EntityClasses
Imports Cito.Tester.ContentModel
Imports Cito.Tester.Common
Imports System.Diagnostics
Imports Questify.Builder.Logic.ContentModel.Scoring
Imports System.Xml.Serialization

Public Module ExtensionMethods
    <Extension()>
    Public Function Deserialize(Of T As New)(ByVal xml As System.Xml.Serialization.XmlSerializer, ByVal data As String) _
        As T

        Dim ret As T
        Using strm As New System.IO.StringReader(data)
            ret = DirectCast(xml.Deserialize(strm), T)
        End Using

        Return ret
    End Function

    Public Function Deserialize(Of T)(input As XElement) As T
        Dim ret As T
        Dim s = New XmlSerializer(GetType(T))

        Using m As New StringReader(input.ToString())
            ret = DirectCast(s.Deserialize(m), T)
        End Using

        Return ret
    End Function

    <Extension()>
    Function DoSerialize(Of T)(obj As T) As XElement
        Dim s = New XmlSerializer(GetType(T))
        Dim ret As XElement = Nothing
        Using m As New StringWriter()
            s.Serialize(m, obj)
            ret = XElement.Parse(m.ToString())
        End Using
        Return ret
    End Function

    <Extension()>
    Public Function Serialize(Of T)(ByVal xml As XmlSerializer, ByVal data As T) As String

        Dim ret As String
        Using strm As New StringWriter
            xml.Serialize(strm, data)
            ret = strm.ToString()
        End Using
        Return ret
    End Function

    <Extension()>
    Function ToXmlNode(ByVal element As XElement) As XmlNode
        Using xmlReader As XmlReader = element.CreateReader()
            Dim xmlDoc As New XmlDocument()
            xmlDoc.Load(xmlReader)
            Return xmlDoc.FirstChild
        End Using
    End Function

    <Extension>
    Public Function SetXmlAsBinData(resource As ResourceEntity, data As XElement) As Boolean
        If (resource.ResourceData Is Nothing) Then resource.ResourceData = New ResourceDataEntity With {.FileExtension = ".xml"}

        Using stream = New MemoryStream()
            SerializeHelper.XmlSerializeToStream(stream, data)
            resource.ResourceData.BinData = stream.ToArray()
        End Using
        Return True
    End Function

    <Extension>
    Public function SetAssessmentToAssessmentResource(resource As AssessmentTestResourceEntity, assessmentTest As AssessmentTest2) As Boolean
        If (resource.ResourceData Is Nothing) Then
            resource.ResourceData = New ResourceDataEntity With {.FileExtension = ".xml"}
        End If
        resource.ResourceData.BinData = SerializeHelper.XmlSerializeToByteArray(assessmentTest)
        Return True
    End function

    <Extension>
    Public Function AddSubParameters(Of TScorePrm As ScoringParameter)(scoreParam As TScorePrm,
                                                                        ParamArray ids As String()) As TScorePrm
        scoreParam.Value = New ParameterSetCollection()
        For Each id In ids
            scoreParam.Value.Add(New ParameterCollection() With {.Id = id})
        Next
        Return scoreParam
    End Function

    <Extension>
    Public Function AsCombinedScoringMap(scoringParameter As ScoringParameter, ParamArray setNumbers As Integer()) _
        As CombinedScoringMapKey
        Dim ret = New List(Of ScoringMapKey)()
        For Each k In scoringParameter.Value
            ret.Add(New ScoringMapKey(scoringParameter, k.Id))
        Next
        Return CombinedScoringMapKey.Create(ret, setNumbers)
    End Function

    <Extension>
    Public Function AsCombinedScoringMap(scoringParameter As ScoringParameter, solution As Solution) _
        As CombinedScoringMapKey
        Return New ScoringMap(New ScoringParameter() {scoringParameter}, solution).GetMap().First()
    End Function

    <Extension>
    Public Function AsScoringMapKey(scoringParameter As ScoringParameter, factId As String) As ScoringMapKey
        Debug.Assert(scoringParameter.Value.Any(Function(subParamSet) subParamSet.Id = factId),
                     "FactId was not found!?!?")
        Return New ScoringMapKey(scoringParameter, factId)
    End Function

    <Extension>
    Public Function [To](Of T)(input As XElement) As T
        Dim ret As T
        Dim s = New XmlSerializer(GetType(T))

        Using m As New StringReader(input.ToString())
            ret = DirectCast(s.Deserialize(m), T)
        End Using

        Return ret
    End Function

    <Extension>
    Public Sub WriteToDebug(s As Solution, stateName As String)
        Dim a As New XmlSerializer(GetType(Solution))
        Debug.WriteLine(String.Empty)
        Debug.WriteLine(String.Format("WriteSolution for State [{0}]", stateName))
        Using stream = New StringWriter()
            a.Serialize(stream, s)

            Debug.WriteLine(stream.ToString())
        End Using
    End Sub
End Module